### PR TITLE
[CAMEL-19620] - upgrade californium-scandium

### PR DESCRIPTION
**Pull request by kahboom**
# Description

This PR intends to upgrade Californium-Scandium from `2.8.0` to `3.10.0` ([CAMEL-19620](https://issues.apache.org/jira/browse/CAMEL-19620)), which also seems to require upgrading Californium to `3.10.0`. Part of this was following the [migration hints](https://github.com/eclipse-californium/californium/blob/main/MIGRATION_HINTS.md) and then searching for replacements of deprecated methods.

Some were relatively straightforward, namely these:
- `PskStore`> `AdvancedPskStore`
- `setRpkTrustStore` > `setAdvancedCertificateVerifier`
- `TrustedRpkStore` > `NewAdvancedCertificateVerifier`

However, I've not been able to fix the tests despite updating the routes to use the new parameters, failing in the `CamelContext` bean binding. Any help would be greatly appreciated.

# Target

- [x] I checked that the commit is targeting the correct branch (note that Camel 3 uses `camel-3.x`, whereas Camel 4 uses the `main` branch)

# Tracking
- [x] If this is a large change, bug fix, or code improvement, I checked there is a [JIRA issue](https://issues.apache.org/jira/browse/CAMEL) filed for the change (usually before you start working on it).

[CAMEL-19620](https://issues.apache.org/jira/browse/CAMEL-19620)

# Apache Camel coding standards and style

- [x] I checked that each commit in the pull request has a meaningful subject line and body.

- [x] I have run `mvn clean install -DskipTests` locally and I have committed all auto-generated changes

Yes there's a bunch of generated stuff, I hope I did it right..

### CODE REVIEW

**Comment by boaks** [comment]
Hi @kahboom,

I answered your e-mail on the [cf-dev-mailing-list](https://www.eclipse.org/lists/cf-dev/msg01795.html).
Is it OK to comment here?
Is it possible to run the test local as well? If so, are there instruction?

**Comment by kahboom** [comment]
Hi @boaks - Yes it's ok to comment here, thank you for your help.

Locally you first build with `mvn clean install -Dquickly` though I think that's not even necessary for this PR (@apupier can maybe confirm).

When you build this component it will run the tests: run `mvn clean verify` from the `camel/components/camel-coap` directory

File: components/camel-coap/src/generated/resources/org/apache/camel/coap/coap+tcp.json [hunk 1]
```diff
@@ -44,16 +44,16 @@
...
-    "trustedRpkStore": { "index": 17, "kind": "parameter", "displayName": "Trusted Rpk Store", "group": "security", "label": "security", "required": false, "type": "object", "javaType": "org.eclipse.californium.scandium.dtls.rpkstore.TrustedRpkStore", "deprecated": false, "autowired": false, "secret": false, "description": "Set the TrustedRpkStore to use to determine trust in raw public keys." }
+    "advancedCertificateVerifier": { "index": 9, "kind": "parameter", "displayName": "Advanced Certificate Verifier", "group": "security", "label": "security", "required": false, "type": "object", "javaType": "org.eclipse.californium.scandium.dtls.x509.NewAdvancedCertificateVerifier", "deprecated": false, "autowired": false, "secret": false, "description": "Set the AdvancedCertificateVerifier to use to determine trust in raw public keys." },
```
Comment (boaks) [inline components/camel-coap/src/generated/resources/org/apache/camel/coap/coap+tcp.json:49]: TCP/TLS support is still "experimental". It doesn't share too much with the DTLS possibilities. The configuration properties are defined in [TcpConfig](https://github.com/eclipse-californium/californium/blob/main/element-connector/src/main/java/org/eclipse/californium/elements/config/TcpConfig.java).

The `AdvancedCertificateVerifier` is only supported for DTLS. 
For TLS the standard java SSLContext is used. As long as you don't want to use remote-calls to fetch/verify certificates, that works pretty well and is very common for java.

File: components/camel-coap/src/generated/resources/org/apache/camel/coap/coap+tcp.json [hunk 2]
```diff
@@ -44,16 +44,16 @@
...
+    "advancedPskStore": { "index": 10, "kind": "parameter", "displayName": "Advanced Psk Store", "group": "security", "label": "security", "required": false, "type": "object", "javaType": "org.eclipse.californium.scandium.dtls.pskstore.AdvancedPskStore", "deprecated": false, "autowired": false, "secret": false, "description": "Set the AdvancedPskStore to use for pre-shared key." },
```
Comment (boaks) [inline components/camel-coap/src/generated/resources/org/apache/camel/coap/coap+tcp.json:50]: Californium doesn't support PSK for TLS.

File: components/camel-coap/src/generated/resources/org/apache/camel/coap/coap+tcp.json [hunk 3]
```diff
@@ -44,16 +44,16 @@
...
+    "cipherSuites": { "index": 12, "kind": "parameter", "displayName": "Cipher Suites", "group": "security", "label": "security", "required": false, "type": "string", "javaType": "java.lang.String", "deprecated": false, "autowired": false, "secret": false, "description": "Sets the cipherSuites String. This is a comma separated String of ciphersuites to configure. If it is not specified, then it falls back to getting the ciphersuites from the sslContextParameters object." },
```
Comment (boaks) [inline components/camel-coap/src/generated/resources/org/apache/camel/coap/coap+tcp.json:52]: The supported "cipher suite" are passed in with the SSLContext/SSLParameters, as common in java.

File: components/camel-coap/src/generated/resources/org/apache/camel/coap/coap+tcp.json [hunk 4]
```diff
@@ -44,16 +44,16 @@
...
+    "clientAuthentication": { "index": 13, "kind": "parameter", "displayName": "Client Authentication", "group": "security", "label": "security", "required": false, "type": "string", "javaType": "java.lang.String", "deprecated": false, "autowired": false, "secret": false, "description": "Sets the configuration options for server-side client-authentication requirements. The value must be one of NONE, WANT, REQUIRE. If this value is not specified, then it falls back to checking the sslContextParameters.getServerParameters().getClientAuthentication() value." },
```
Comment (boaks) [inline components/camel-coap/src/generated/resources/org/apache/camel/coap/coap+tcp.json:53]: I'm not sure, how this "json" works. Just to say:

[TcpConfig.TLS_CLIENT_AUTHENTICATION_MODE](https://github.com/eclipse-californium/californium/blob/main/element-connector/src/main/java/org/eclipse/californium/elements/config/TcpConfig.java#L82-L87) uses an enumeration [CertificateAuthenticationMode](https://github.com/eclipse-californium/californium/blob/main/element-connector/src/main/java/org/eclipse/californium/elements/config/CertificateAuthenticationMode.java), not strings.

File: components/camel-coap/src/generated/resources/org/apache/camel/coap/coap+tcp.json [hunk 5]
```diff
@@ -44,16 +44,16 @@
...
+    "privateKey": { "index": 14, "kind": "parameter", "displayName": "Private Key", "group": "security", "label": "security", "required": false, "type": "object", "javaType": "java.security.PrivateKey", "deprecated": false, "autowired": false, "secret": true, "description": "Set the configured private key for use with Raw Public Key." },
+    "publicKey": { "index": 15, "kind": "parameter", "displayName": "Public Key", "group": "security", "label": "security", "required": false, "type": "object", "javaType": "java.security.PublicKey", "deprecated": false, "autowired": false, "secret": false, "description": "Set the configured public key for use with Raw Public Key." },
+    "recommendedCipherSuitesOnly": { "index": 16, "kind": "parameter", "displayName": "Recommended Cipher Suites Only", "group": "security", "label": "security", "required": false, "type": "boolean", "javaType": "boolean", "deprecated": false, "autowired": false, "secret": false, "defaultValue": true, "description": "The CBC cipher suites are not recommended. If you want to use them, you first need to set the recommendedCipherSuitesOnly option to false." },
```
Comment (boaks) [inline components/camel-coap/src/generated/resources/org/apache/camel/coap/coap+tcp.json:56]: only via SSLContext.

File: components/camel-coap/src/main/java/org/apache/camel/coap/CoAPComponent.java [hunk 6]
```diff
@@ -68,34 +70,30 @@ public synchronized CoapServer getServer(int port, CoAPEndpoint endpoint) throws
...
                 TcpServerConnector tcpConnector = null;
                 // TLS + TCP
                 if (endpoint.getUri().getScheme().startsWith("coaps")) {
-                    int tlsHandshakeTimeout = config.getInt(NetworkConfig.Keys.TLS_HANDSHAKE_TIMEOUT);
-
                     SSLContext sslContext = endpoint.getSslContextParameters().createSSLContext(getCamelContext());
-                    TlsServerConnector.ClientAuthMode clientAuthMode = TlsServerConnector.ClientAuthMode.NONE;
                     if (endpoint.isClientAuthenticationRequired()) {
-                        clientAuthMode = TlsServerConnector.ClientAuthMode.NEEDED;
+                        config.set(DtlsConfig.DTLS_CLIENT_AUTHENTICATION_MODE, CertificateAuthenticationMode.NEEDED);
```
Comment (boaks) [inline components/camel-coap/src/main/java/org/apache/camel/coap/CoAPComponent.java:87]: As I wrote above, for TCP/TLS the configuration properties are defined in [TcpConfig](https://github.com/eclipse-californium/californium/blob/main/element-connector/src/main/java/org/eclipse/californium/elements/config/TcpConfig.java). Using [DtlsConfig](https://github.com/eclipse-californium/californium/blob/main/scandium-core/src/main/java/org/eclipse/californium/scandium/config/DtlsConfig.java) applies only to DTLS.

File: components/camel-coap/src/main/java/org/apache/camel/coap/CoAPEndpoint.java [hunk 7]
```diff
@@ -405,19 +420,20 @@ public static boolean enableTCP(URI uri) {
 
     public DTLSConnector createDTLSConnector(InetSocketAddress address, boolean client) throws IOException {
 
-        DtlsConnectorConfig.Builder builder = new DtlsConnectorConfig.Builder();
+        DtlsConnectorConfig.Builder builder = new DtlsConnectorConfig.Builder(Configuration.getStandard());
+        CertificateConfigurationHelper certificateConfigurationHelper = new CertificateConfigurationHelper();
```
Comment (boaks) [inline components/camel-coap/src/main/java/org/apache/camel/coap/CoAPEndpoint.java:424]: That's not generally needed. Usually the DtlsConnectorConfig.Builder creates one automatically and applies that to the certificate provider and verifier. The idea behind this is, if remote implementations are used, traversing the credentials may fail and so someone can pass in the default values according the requirements. If local "stores" are used, then the default traverse them and collect the default parameters automatically. 
I would remove it here. 
Otherwise, if you want to keep it, you would need to apply it to the certificate provider and verifier in order to get the defaults filled (see ConfigurationHelperSetup).

File: components/camel-coap/src/main/java/org/apache/camel/coap/CoAPEndpoint.java [hunk 8]
```diff
@@ -460,31 +482,40 @@ public DTLSConnector createDTLSConnector(InetSocketAddress address, boolean clie
...
+                        .builder()
+                        .setTrustedCertificates(certs)
+                        .build();
+                builder.setAdvancedCertificateVerifier(trust);
             }
-            if (trustedRpkStore != null) {
-                builder.setTrustCertificateTypes(CertificateType.RAW_PUBLIC_KEY);
-                builder.setRpkTrustStore(trustedRpkStore);
+            if (advancedCertificateVerifier != null) {
+                builder.set(DTLS_CERTIFICATE_TYPES, Arrays.asList(CertificateType.RAW_PUBLIC_KEY));
```
Comment (boaks) [inline components/camel-coap/src/main/java/org/apache/camel/coap/CoAPEndpoint.java:508]: Using Raw Public Key requires to set the trusted keys explicitly in the certificate verifier. x509 certificates are not considered implicit as Raw Public Key trust.

File: components/camel-coap/src/main/java/org/apache/camel/coap/CoAPEndpoint.java [hunk 9]
```diff
@@ -500,18 +531,14 @@ public CoapClient createCoapClient(URI uri) throws IOException, GeneralSecurityE
...
-            int tcpThreads = config.getInt(NetworkConfig.Keys.TCP_WORKER_THREADS);
-            int tcpConnectTimeout = config.getInt(NetworkConfig.Keys.TCP_CONNECT_TIMEOUT);
-            int tcpIdleTimeout = config.getInt(NetworkConfig.Keys.TCP_CONNECTION_IDLE_TIMEOUT);
             TcpClientConnector tcpConnector = null;
 
             // TLS + TCP
             if (getUri().getScheme().startsWith("coaps")) {
                 SSLContext sslContext = getSslContextParameters().createSSLContext(getCamelContext());
-                tcpConnector = new TlsClientConnector(sslContext, tcpThreads, tcpConnectTimeout, tcpIdleTimeout);
+                tcpConnector = new TlsClientConnector(sslContext, Configuration.createStandardWithoutFile());
```
Comment (boaks) [inline components/camel-coap/src/main/java/org/apache/camel/coap/CoAPEndpoint.java:539]: Using `Configuration.createStandardWithoutFile()` will use the defaults only.
There is a difference of calling "createStandardWithoutFile" and "getStandard" successive. Usually Californium uses a property file for the settings. If you don't want that or the environment don't support files, calling "createStandardWithoutFile" overcomes that. The you usually apply your config values and the use that Configuration instance for building new Californium components. In same cases (historically), users call createStandardWithoutFile, apply their changes and the call "getStandard()" instead of passing the Configuration instance around.  Some may even use createStandardWithoutFile by intention, but then it's required to adapt the configuration again (or take the defaults).
Hope it gets clearer.

**Comment by boaks** [comment]
Lets see, if the the comments help to reduce the failures ;-).

**Comment by kahboom** [comment]
Wow thank you @boaks , that's very helpful. Working through your feedback now.

**Review by oscerd — APPROVED** [review]

**Comment by boaks** [comment]
I simply don't see in the logs, what's failing. 
Does someone have more insights?

**Comment by oscerd** [comment]
There are uncommitted changes. The branch needs to be rebuilt.

File: components/camel-coap/src/generated/resources/org/apache/camel/coap/coap+tcp.json (hunk 1 above)
Comment (kahboom) [reply components/camel-coap/src/generated/resources/org/apache/camel/coap/coap+tcp.json:49]: Thanks @boaks . The JSON files and `CoapEndpointConfigurer` are automatically generated.

From what I can see `AdvancedCertificateVerifier` and `PSK` are only used in the DTLS configuration setup ([here](https://github.com/apache/camel/pull/12769/files#diff-fa9b8056c9aaac49ce4368ccadc8e7d9748f3dbdb1d0a59481b8cecdbf8d6cc7R424)).

**Comment by davsclaus** [comment]
These changed files needs to be in this PR 

	modified:   catalog/camel-catalog/src/generated/resources/org/apache/camel/catalog/components/coap+tcp.json
	modified:   catalog/camel-catalog/src/generated/resources/org/apache/camel/catalog/components/coap.json
	modified:   catalog/camel-catalog/src/generated/resources/org/apache/camel/catalog/components/coaps+tcp.json
	modified:   catalog/camel-catalog/src/generated/resources/org/apache/camel/catalog/components/coaps.json

**Comment by davsclaus** [comment]
[ERROR] Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.033 s <<< FAILURE! -- in org.apache.camel.coap.CoAPMethodRestrictTest
[ERROR] org.apache.camel.coap.CoAPMethodRestrictTest.testSpecifiedCoAPMethodRestrict -- Time elapsed: 0.014 s <<< FAILURE!
org.opentest4j.AssertionFailedError: expected: <null> but was: <>
	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
	at org.junit.jupiter.api.AssertNull.failNotNull(AssertNull.java:50)
	at org.junit.jupiter.api.AssertNull.assertNull(AssertNull.java:35)
	at org.junit.jupiter.api.AssertNull.assertNull(AssertNull.java:30)
	at org.junit.jupiter.api.Assertions.assertNull(Assertions.java:279)
	at org.apache.camel.coap.CoAPMethodRestrictTest.assertCoAPMethodRestrictResponse(CoAPMethodRestrictTest.java:61)
	at org.apache.camel.coap.CoAPMethodRestrictTest.testSpecifiedCoAPMethodRestrict(CoAPMethodRestrictTest.java:42)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)

**Comment by boaks** [comment]
I'm not common to your test-code. So for me it's hard to see, what this actually tested. 
From the failure message:

```
expected: but was: <>
...
at org.apache.camel.coap.CoAPMethodRestrictTest.assertCoAPMethodRestrictResponse(CoAPMethodRestrictTest.java:61)
```

Points to:

```
assertNull(result);
```

One guess, maybe a wrong one:

[MIGRATION_HINTS](https://github.com/eclipse-californium/californium/blob/main/MIGRATION_HINTS.md#californium-core)

```
Message.getPayload():

Since 3.0 null is replaced by Bytes.EMPTY. The method will now always return an byte array, which may be empty.
```

File: components/camel-coap/src/test/java/org/apache/camel/coap/CoAPComponentTLSTestBase.java [hunk 10]
```diff
@@ -207,9 +208,9 @@ public void configure() {
                         PORT2)
                         .transform(body().prepend("Hello "));
 
-                fromF(getProtocol()
-                      + "://localhost:%d/TestResource?sslContextParameters=#clientAuthServiceSSLContextParameters", PORT3)
-                        .transform(body().prepend("Hello "));
+                //                fromF(getProtocol()
+                //                      + "://localhost:%d/TestResource?sslContextParameters=#clientAuthServiceSSLContextParameters", PORT3)
+                //                        .transform(body().prepend("Hello "));
```
Comment (apupier) [inline components/camel-coap/src/test/java/org/apache/camel/coap/CoAPComponentTLSTestBase.java:213]: commented out code to remove, or to readd with correct change?

File: components/camel-coap/src/test/java/org/apache/camel/coap/CoAPComponentTLSTestBase.java [hunk 11]
```diff
@@ -399,14 +405,21 @@ private void registerTLSConfiguration(CamelContext context) throws GeneralSecuri
...
-        };
-        TrustedRpkStore failedTrustedRpkStore = id -> {
-            return false;
-        };
+        //        TrustedRpkStore trustedRpkStore = id -> {
+        //            return true;
+        //        };
+        //        TrustedRpkStore failedTrustedRpkStore = id -> {
+        //            return false;
+        //        };
```
Comment (apupier) [inline components/camel-coap/src/test/java/org/apache/camel/coap/CoAPComponentTLSTestBase.java:413]: commented out code to remove

File: components/camel-coap/src/test/java/org/apache/camel/coap/CoAPComponentTLSTestBase.java [hunk 12]
```diff
@@ -419,9 +432,10 @@ private void registerTLSConfiguration(CamelContext context) throws GeneralSecuri
 
         context.getRegistry().bind("privateKey", privateKey);
         context.getRegistry().bind("publicKey", publicKey);
-        context.getRegistry().bind("trustedRpkStore", trustedRpkStore);
-        context.getRegistry().bind("failedTrustedRpkStore", failedTrustedRpkStore);
-        context.getRegistry().bind("pskStore", pskStore);
+
+        context.getRegistry().bind("advancedCertificateVerifier", advancedCertificateVerifier);
+        //        context.getRegistry().bind("failedAdvancedCertificateVerifier", failedAdvancedCertificateVerifier);
```
Comment (apupier) [inline components/camel-coap/src/test/java/org/apache/camel/coap/CoAPComponentTLSTestBase.java:437]: commented out code to remove ro to readd correctly, I would expect some tests to fail without it

Comment (kahboom) [reply components/camel-coap/src/test/java/org/apache/camel/coap/CoAPComponentTLSTestBase.java:437]: Thanks, should have caught these!

**Comment by kahboom** [comment]
>Since 3.0 null is replaced by `Bytes.EMPTY`. The method will now always return an byte array, which may be empty.

@boaks thanks for the tip, I've changed it from `assertNull(result)` to `assertArrayEquals(Bytes.EMPTY, result.getBytes());` which seems to have fixed the error.

Apologies for the slow progress on this PR, I'm a novice. :)

These are the remaining test failures I'm trying to work through (plus the system out logging that Claus mentioned):

```
CoAPComponentTCPTLSTest>CoAPComponentTLSTestBase.testCall:73 mock://result Received message count. Expected: <1> but was: <0>

CoAPComponentTLSTest>CoAPComponentTLSTestBase.testCall:73 mock://result Received message count. Expected: <1> but was: <0>

CoAPComponentTLSTest>CoAPComponentTLSTestBase.testPreSharedKeyCipherSuite:170 mock://result Received message count. Expected: <1> but was: <0>

CoAPObserveTest.testClientAndServerObservable:80 mock://sourceResults Received message count. Expected: <2> but was: <1>
```

**Comment by davsclaus** [comment]
Welcome to the Camel github community, - your first PR is certainly not an easy one, so anyone of us would have trouble getting this done without help.

You are welcome to reach out to us on zulip chat if you are looking for other and maybe more easier and quicker PRs you can do within a single day.

**Comment by kahboom** [comment]
I found there was a duplicated `advancedPsk` URL param trailing at the end of the test route, which resolves another one of the errors.

So now I'm down to two, which seem to be the same:

```
CoAPComponentTCPTLSTest>CoAPComponentTLSTestBase.testCall:73 mock://result Received message count. Expected: <1> but was: <0>

CoAPComponentTLSTest>CoAPComponentTLSTestBase.testCall:73 mock://result Received message count. Expected: <1> but was: <0>
```

**Comment by boaks** [comment]
> So now I'm down to two

I still didn't clone and test it on my own locally. Therefore, if possible, may be someone can provide the logs or some details about the last two left failures.

**Comment by kahboom** [comment]
Sorry about that @boaks :

```
[ERROR] Tests run: 15, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 10.83 s <<< FAILURE! -- in org.apache.camel.coap.CoAPComponentTCPTLSTest
[ERROR] org.apache.camel.coap.CoAPComponentTCPTLSTest.testCall(String)[3] -- Time elapsed: 10.07 s <<< FAILURE!
java.lang.AssertionError: mock://result Received message count. Expected: <1> but was: <0>
	at org.apache.camel.component.mock.MockEndpoint.fail(MockEndpoint.java:1952)
	at org.apache.camel.component.mock.MockEndpoint.assertEquals(MockEndpoint.java:1886)
	at org.apache.camel.component.mock.MockEndpoint.doAssertIsSatisfied(MockEndpoint.java:449)
	at org.apache.camel.component.mock.MockEndpoint.assertIsSatisfied(MockEndpoint.java:427)
	at org.apache.camel.component.mock.MockEndpoint.assertIsSatisfied(MockEndpoint.java:417)
	at org.apache.camel.component.mock.MockEndpoint.assertIsSatisfied(MockEndpoint.java:213)
	at org.apache.camel.coap.CoAPComponentTLSTestBase.testCall(CoAPComponentTLSTestBase.java:73)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
```

And the second:

```
[ERROR] Tests run: 15, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 164.9 s <<< FAILURE! -- in org.apache.camel.coap.CoAPComponentTLSTest
[ERROR] org.apache.camel.coap.CoAPComponentTLSTest.testCall(String)[3] -- Time elapsed: 72.15 s <<< FAILURE!
java.lang.AssertionError: mock://result Received message count. Expected: <1> but was: <0>
	at org.apache.camel.component.mock.MockEndpoint.fail(MockEndpoint.java:1952)
	at org.apache.camel.component.mock.MockEndpoint.assertEquals(MockEndpoint.java:1886)
	at org.apache.camel.component.mock.MockEndpoint.doAssertIsSatisfied(MockEndpoint.java:449)
	at org.apache.camel.component.mock.MockEndpoint.assertIsSatisfied(MockEndpoint.java:427)
	at org.apache.camel.component.mock.MockEndpoint.assertIsSatisfied(MockEndpoint.java:417)
	at org.apache.camel.component.mock.MockEndpoint.assertIsSatisfied(MockEndpoint.java:213)
	at org.apache.camel.coap.CoAPComponentTLSTestBase.testCall(CoAPComponentTLSTestBase.java:73)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
```

Here is the full output: https://gist.github.com/kahboom/9f5b645f0c2b92f7f859b607ecbac2f4

I wonder if it's something to do with the SSL parameters in `CoAPComponentTLSTestBase.java` (`clientSSLContextParameters` and `clientAuthClientSSLContextParameters`).

- This is the test: [CoAPComponentTLSTestBase.java#L73](https://github.com/apache/camel/blob/5a3524e645de92dbb153e2202a58692601459c2a/components/camel-coap/src/test/java/org/apache/camel/coap/CoAPComponentTLSTestBase.java#L73)
- The test routes: [CoAPComponentTLSTestBase.java#L216](https://github.com/apache/camel/blob/5a3524e645de92dbb153e2202a58692601459c2a/components/camel-coap/src/test/java/org/apache/camel/coap/CoAPComponentTLSTestBase.java#L216) (`"direct:start"`, `"direct:selfsigned"`, `"direct:clientauth"`, `"direct:ciphersuites"`)
- The SSL parameters it binds to [here](https://github.com/apache/camel/blob/5a3524e645de92dbb153e2202a58692601459c2a/components/camel-coap/src/test/java/org/apache/camel/coap/CoAPComponentTLSTestBase.java#L362): `clientSSLContextParameters` and `clientAuthClientSSLContextParameters`

**Comment by davsclaus** [comment]
There are again changed files in catalog/camel-catalog that are not in this PR so the CI cannot run the tests

**Comment by kahboom** [comment]
@davsclaus is it fine now? I rebased, ran `mvn clean verify` and then regenerated with `mvn org.apache.camel:camel-package-maven-plugin:generate`

**Comment by boaks** [comment]
> The test routes: [CoAPComponentTLSTestBase.java#L216](https://github.com/apache/camel/blob/5a3524e645de92dbb153e2202a58692601459c2a/components/camel-coap/src/test/java/org/apache/camel/coap/CoAPComponentTLSTestBase.java#L216) ("direct:start", "direct:selfsigned", "direct:clientauth", "direct:ciphersuites")

I'm not common to this test system.
Which side is using the "selfsigned certificate"? I guess the client.
How is the server intended  to trust that? Is that "client-selfsigned-certificate" added to the server's trust? 

Some Californium details about that trust handling:
For TLS Californium uses a "TrustManager". The JCE implementation requires then the "selfsigned certificate" to be added in order to trust it. For DTLS, more or less for testing and historical reasons, we use `null` for "trust none", and `new X509Certificate[0]` for "trust all". But if you pass that `new X509Certificate[0]` to the standard java JCE TrustManager, I guess this will be "trust none". In order to overcome that, again only for tests, Californium offers "SslContextUtil.createTrustAllManager()". 

Anyway, I guess, before mixing the "test only" approach into real code, it's better to ensure, that the "selfsigned certifcate" are added to the trust of the other side.

**Comment by davsclaus** [comment]
Trying this PR from a fresh start at
https://github.com/apache/camel/pull/13039

Want to make the CI able to build and test so we can better finish this work

**Comment by davsclaus** [comment]
closing this PR as JIRA has a link to it from
https://issues.apache.org/jira/browse/CAMEL-20395

**Comment by kahboom** [comment]
thanks @davsclaus , apologies I was unable to finish this. perhaps I (or someone else) can try picking it up again in the future.

**Comment by davsclaus** [comment]
> thanks @davsclaus , apologies I was unable to finish this. perhaps I (or someone else) can try picking it up again in the future.

Absolutely no problem, you got 99% there. We only have that one weird test - we have a JIRA and will look at this later.
